src/cargo/ops/cargo_rustc/context.rs

```diff
@@ -483,6 +483,15 @@ impl<'a, 'cfg> Context<'a, 'cfg> {
         // when changing feature sets each lib is separately cached.
         self.resolve.features_sorted(unit.pkg.package_id()).hash(&mut hasher);
 
+        // Mix in the target-metadata of all the dependencies of this target
+        if let Ok(deps) = self.used_deps(unit) {
+            let mut deps_metadata = deps.into_iter().map(|dep_unit| {
+                self.target_metadata(&dep_unit)
```

Comment:
So runtime-wise we've had a lot of issues in the past with this sort of recursive calculation causing an exponential amount of code to run instead of linear (wrt the dependency graph). For example here I think that if we call `target_metadata` for all targets this goes from linear (currently) to exponential (after this PR), right?

Perhaps we can introduce a cache for `target_metadata`? (that's what we do for everything else!)

This tends to not show up much in small crate graphs but projects like Servo (and in theory Dropbox w/ 200+ crates) may end up showing this pretty badly.

Reply:
Yeah. That sounds like a great solution. Envisioning a big hashmap from Unit -> Metadata in the ctx? We could probably even precalculate it if we walk the units in dependency order.

Reply:
Sounds great! I'm fine with doing the caching lazily or doing it all at once when we walk in dependency order. I know we have a few "prepopulate passes" at the beginning to walk the graph, but I sort of forget what they're doing. Basically whatever's easiest is fine by me